### PR TITLE
Pid

### PR DESCRIPTION
**Pull request by vkurpad**
## Purpose
Added new skill to parse engineering drawings

## Does this introduce a breaking change?

```
[ ] Yes
[ X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
Add skill to skillset and test wit an engineering drawing

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: Vision/PID/README.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+- azure-functions
+name: PID 
+urlFragment: azure-pid-drawing-sample
+description: This custom skill extracts specific product/equipment names from PI&D drawings.
+Deploy: Manually [deploy the container image as a Azure function](#deployment)
+---
+
+# PID Skill
+
+P&ID is a graphical representation of the actual process plant using various symbols that represent actual equipment. This skill is designed to extract equipment infomration  from specific instrument symbols in engineering diagrams. The skill uses the X, Y coordinates of text extracted by OCR to generate groupings of text based on proximity, vertical and horizontal seperation and alignment.
```
Comment (bleroy) [inline Vision/PID/README.md:33]: You might want to explain what P&ID means, and put a link to https://en.wikipedia.org/wiki/Piping_and_instrumentation_diagram

File: Vision/PID/README.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+![diagram](/images/custom_skill_design.png).
+
+## Requirements
+
+This skill requires Docker to build a container that will be deployed as an Azure function.
+
+## Settings
+
+The default configuration of the skill identified tags or equipment and associated text blocks. Tuning the following parameters allos you to se the sensitivity of grouping of individual text spans into a block.
```
Comment (bleroy) [inline Vision/PID/README.md:45]: identified -> identifies
allos -> allows

File: Vision/PID/README.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+1. ```maxSegment``` defines the max length of a valid text segment
+2. ```leftAlignSensitivity``` defines the sensitivity of the algorithm in matching text blocks that are left aligned
+3. ```rightAlignSensitivity``` defines the sensitivity of the algorithm in matching text blocks that are right aligned
+4. ```centerAlignSensitivty``` defines the sensitivity of the algorithm in matching text blocks that are center aligned
+
+## Deployment
+
+Follow these steps to buld the container and deploy the skill as an Azure Function.
```
Comment (bleroy) [inline Vision/PID/README.md:54]: buld -> build

File: Vision/PID/README.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+
+1. Navigate to the diagramskill folder and build the docker container ```docker build -t pidskill .```
```
Comment (bleroy) [inline Vision/PID/README.md:57]: diagramskill -> diagram skill (I hope it's not diagrams kill :D )

File: Vision/PID/README.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+2. Run the container ```docker run -p 8080:80 -it pidskill:latest```
+3. Save the image ```docker commit {container id from previous step} pidskill```
+4. Push the image to the container registry ```docker push {containerregistry}.azurecr.io/pidskill```
+
+Once the image is in the container registry, you can now create anAzure function to deploy that image to.
```
Comment (bleroy) [inline Vision/PID/README.md:62]: anAzure -> an Azure

File: Vision/PID/README.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+1. In the portal, create a new Azure Function App
+2. Select the Docker Container option, provide a valid function name
+3. Once the deployment is complete, navigate to the resource, select ```Container settings```
+4. Select Azure Container Registry for the Image Source 
+5. Select the registry, image and tag
+6. Select continuous deployment to On to ensure that the skill is updated when a new image is uploaded
```
Comment (bleroy) [inline Vision/PID/README.md:69]: Select -> Set

File: Vision/PID/diagramskill/app/__init__.py [hunk 7]
```diff
@@ -0,0 +1,491 @@
...
+    maxSegment = 95
+    leftAlignSensitivity = 5
+    rightAlignSensitivity = 5
+    centerAlignSensitivty = 15
+ 
+    for i in range(len(textblocks)):
+        tb = textblocks[i].copy()
+        count=0
+        for j in range(len(textblocks)):
+            if i != j:
```
Comment (bleroy) [inline Vision/PID/diagramskill/app/__init__.py:291]: Those nested ifs look like they should be AND'ed

File: Vision/PID/diagramskill/debug.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import cv2 
+from azure.storage.blob import BlobServiceClient
+import numpy as np
+
+def write_with_bounding(img,boxes):
+    color = (200,0,0)
+    for b in boxes:  
+        cv2.rectangle(img, (b.topleft_x, b.topleft_y), (b.bottomright_x,b.bottomright_y), color, 3)
+
+    #if circles is not None:
```
Comment (bleroy) [inline Vision/PID/diagramskill/debug.py:12]: Why is this commented out?

File: Vision/PID/README.md (hunk 4 above)
Comment (vkurpad) [reply Vision/PID/README.md:57]: The folder name is a single token diagramskill

File: Vision/PID/diagramskill/debug.py (hunk 8 above)
Comment (vkurpad) [reply Vision/PID/diagramskill/debug.py:12]: This is only for testing and is not needed by customers. I just left it in if someone wants to modify the code and needed the scaffolding

File: Vision/PID/README.md (hunk 1 above)
Comment (bleroy) [reply Vision/PID/README.md:33]:
```suggestion
P&ID is a graphical representation of the actual process plant using various symbols that represent actual equipment. This skill is designed to extract equipment information  from specific instrument symbols in engineering diagrams. The skill uses the X, Y coordinates of text extracted by OCR to generate groupings of text based on proximity, vertical and horizontal separation and alignment. 

```

File: Vision/PID/README.md (hunk 4 above)
Comment (bleroy) [reply Vision/PID/README.md:57]: haha, makes sense. Maybe back tick it then:
 
```suggestion
1. Navigate to the `diagramskill` folder and build the docker container ```docker build -t pidskill .```

```

**Comment by bleroy** [comment]
Another thing would be to update the top-level solution and readme, similar to Sorin's changes in #34

File: Vision/PID/diagramskill/app/__init__.py (hunk 7 above)
Comment (vkurpad) [reply Vision/PID/diagramskill/app/__init__.py:291]: works as is... :)

**Comment by vkurpad** [comment]
@bleroy Ready to merge?

File: Vision/PID/diagramskill/app/__init__.py (hunk 7 above)
Comment (bleroy) [reply Vision/PID/diagramskill/app/__init__.py:291]: I'm sure it does, it's just that it would be more readable.

**Pull request merged**